vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java

```diff
@@ -88,23 +95,54 @@
 
     private SerializableConsumer<UI> sizeRequest;
 
+    private CheckboxGroupDataView<T> dataView;
+
     public CheckboxGroup() {
         super(Collections.emptySet(), Collections.emptySet(), JsonArray.class,
                 CheckboxGroup::presentationToModel,
                 CheckboxGroup::modelToPresentation);
         registerValidation();
     }
 
+    @Override
+    public CheckboxGroupDataView<T> setDataSource(
+            DataProvider<T, ?> dataProvider) {
+        setDataProvider(dataProvider);
+        return getDataView();
+    }
 
     @Override
-    public CheckboxGroupListDataView<T> setDataProvider(ListDataProvider<T> dataProvider) {
-        this.setDataProvider((DataProvider<T, ?>) dataProvider);
+    public CheckboxGroupListDataView<T> setDataSource(
+            ListDataProvider<T> dataProvider) {
+        this.setDataProvider(dataProvider);
         return getListDataView();
     }
 
     @Override
     public CheckboxGroupListDataView<T> getListDataView() {
-        return new CheckboxGroupListDataView<>(this::getDataProvider, this);
+        if (getDataProvider() instanceof ListDataProvider) {
+            if (dataView == null || !(dataView instanceof ListDataView)) {
+                dataView = new CheckboxGroupListDataView<>(
+                        this::getDataProvider, this);
+            }
+        }
+        return (CheckboxGroupListDataView) dataView;
```

Comment:
Shouldn't it be inside `if` block and exception thrown for non-list data provider, like it's done for `Select`?

Reply:
Yes it should. Fixed.